tests/unit/test_charm.py

```diff
@@ -168,131 +168,155 @@ def test_on_config_changed(self):
 
+        # Verify that we don't initialise the replica set when number of
+        # expected units doesn't match
+        initialise_replica_set.assert_not_called()
+
+    @patch_network_get(private_address="1.1.1.1")
     @mock.patch("mongoserver.MongoDB.is_ready")
-    @mock.patch("mongoserver.MongoDB.initialize_replica_set")
+    @mock.patch("mongoserver.MongoDB.initialise_replica_set")
     @mock.patch("charm.MongodbOperatorCharm._open_port_tcp")
     @mock.patch("charm.service_resume")
-    def test_on_start_not_leader_doesnt_start_mongo(
-        self, service_resume, _open_port_tcp, initialize_replica_set, is_ready
+    @mock.patch("charm.MongodbOperatorCharm._initialise_replica_set")
+    @mock.patch("charm.MongodbOperatorCharm.app")
+    def test_on_start_not_leader_doesnt_initialise_replica_set(
+        self,
+        app,
+        _initialise_replica_set,
+        service_resume,
+        _open_port_tcp,
+        initialise_replica_set,
+        is_ready,
     ):
         self.harness.set_leader(False)
+        is_ready.return_value = True
         self.harness.charm.on.start.emit()
 
-        _open_port_tcp.assert_not_called()
-        service_resume.assert_not_called()
-        is_ready.assert_not_called()
-        initialize_replica_set.assert_not_called()
-        self.assertNotEqual(self.harness.charm.unit.status, ActiveStatus(""))
+        service_resume.assert_called()
+        _open_port_tcp.assert_called()
+        is_ready.assert_called_with(all_replicas=False)
+        self.assertEqual(self.harness.charm.unit.status, ActiveStatus())
+        app.planned_units.assert_not_called()
+        _initialise_replica_set.assert_not_called()
 
     @patch_network_get(private_address="1.1.1.1")
     @mock.patch("mongoserver.MongoDB.is_ready")
-    @mock.patch("mongoserver.MongoDB.initialize_replica_set")
+    @mock.patch("charm.MongodbOperatorCharm._initialise_replica_set")
     @mock.patch("charm.MongodbOperatorCharm._open_port_tcp")
     @mock.patch("charm.service_resume")
     @mock.patch("charm.service_running")
     def test_on_start_systemd_failure_leads_to_blocked_status(
-        self, service_running, service_resume, _open_port_tcp, initialize_replica_set, is_ready
+        self, service_running, service_resume, _open_port_tcp, initialise_replica_set, is_ready
     ):
-        self.harness.set_leader(True)
-        is_ready.return_value = True
         service_running.return_value = False
         service_resume.return_value = False
 
         with self.assertLogs("charm", "ERROR") as logs:
             self.harness.charm.on.start.emit()
+            service_resume.assert_called()
             self.assertIn("ERROR:charm:failed to enable mongod.service", logs.output)
         self.assertEqual(self.harness.charm.unit.status, BlockedStatus("couldn't start MongoDB"))
-        initialize_replica_set.assert_not_called()
+        _open_port_tcp.assert_not_called()
+        is_ready.assert_not_called()
```

Comment:
It might also be worth checking that initialise_replica_set has not been called.

Reply:
Agreed! Great eye Marcelo